app/controllers/unattended_controller.rb

```diff
@@ -18,13 +18,28 @@ class UnattendedController < ApplicationController
   # all of our requests should be returned in text/plain
   after_action :set_content_type
 
-  # this actions is called by each operatingsystem post/finish script - it notify us that the OS installation is done.
   def built
     logger.info "#{controller_name}: #{@host.name} is Built!"
+    logger.debug "Log lines from the OS installer:"
+    logger.debug(extract_build_log)
+    # clear possible previous errors
+    @host.build_errors = nil
     update_ip if Setting[:update_ip_from_built_request]
     head(@host.built ? :created : :conflict)
   end
 
+  def failed
+    logger.warn "#{controller_name}: #{@host.name} build Failed!"
+    logger.warn "Log lines from the OS installer:"
+    logger.warn(@host.build_errors = extract_build_log)
```

Comment:
Assigning and logging the errors at the same time is quite confusiong, please do this in two lines:

```
@host.build_errors = extract_build_log
logger.warn(@host.build_errors)
```